Fits structural kw docs (#9352)
astropy/astropy: docs/io/fits/index.rst

.. _structural_keywords:

Structural Keywords
"""""""""""""""""""

FITS keywords mix up both metadata and critical information file structure that is
needed to parse the file. These *structural* keywords can be silently modified by
:mod:`astropy.io.fits` and, in general, **should not be touched by the user**.

The specific set of structural keywords used by the FITS standard varies with HDU type.
The following table lists which keywords are associated with each HDU type:

.. csv-table:: Structural Keywords
   :header: "HDU Type", "Structural Keywords"
   :widths: 20, 20

   "All", "``SIMPLE``, ``BITPIX``, ``NAXIS``"
   ":class:`PrimaryHDU`", "``EXTEND``"
   ":class:`ImageHDU`, :class:`TableHDU`, :class:`BinTableHDU`",  "``PCOUNT``, ``GCOUNT``"
   ":class:`GroupsHDU`", "``NAXIS1``, ``GCOUNT``, ``PCOUNT``, ``GROUPS``"
   ":class:`TableHDU`, :class:`BinTableHDU`", "``TFIELDS``, ``TFORM``, ``TBCOL``"

Structural keywords are often checked and/or updated as a consequence of common operations.
For example, when:

1. Reading a file
2. Setting the associated data
3. Setting the :class:`Header` object
4. Writing a file

> 5. Calling an HDU's verify method (e.g., :func:`PrimaryHDU.verify`)

Reviewer: ```suggestion
There are many other reserved keywords, for instance for the data scaling, or for table's column 
attributes, as described in the  `FITS Standard <https://fits.gsfc.nasa.gov/fits_standard.html>`__. 
Most of these are accessible via attributes of the :class:`Column` or HDU objects, for instance 
``hdu.name`` to set ``EXTNAME``, or ``hdu.ver`` for ``EXTVER``. Structural keywords are checked 
and/or updated as a consequence of common operations. For example, when:

1. Setting the data, the ``NAXIS*`` keywords are set from the data shape (``.data.shape``), and ``BITPIX`` 
  from the data type (``.data.dtype``).
2. Setting the header, its keywords are updated based on the data properties (as above). 
3. Writing a file, all the necessary keywords are deleted, updated or added to the header. 
4. Calling an HDU's verify method (e.g., :func:`PrimaryHDU.verify`), some keywords can be fixed automatically. 
```

    See also :ref:`sphx_glr_generated_examples_io_modify-fits-header.py`.

Structural Keywords
"""""""""""""""""""

> .. _structural_keywords:

Reviewer 2: Move this `ref` tag to before the heading, with blank line between them.

    See also :ref:`sphx_glr_generated_examples_io_modify-fits-header.py`.

.. _structural_keywords:

Structural Keywords
"""""""""""""""""""

An unfortunate design quirk of FITS is that FITS keywords mix up both metadata
and critical information file structure that is needed to parse the file.
These *structural* keywords can be silently modified by :mod:`astropy.io.fits` and,
in general, should not be touched by the user.

The specific set of structural keywords used by the FITS standard varies with HDU type.
The following table lists which keywords are associated with each HDU type:

.. csv-table:: Structural Keywords
   :header: "HDU Type", "Structural Keywords"
   :widths: 20, 20

   "All", "``SIMPLE``, ``BITPIX``, ``NAXIS``"
   ":class:`PrimaryHDU`", "``EXTEND``"
   ":class:`ImageHDU`, :class:`TableHDU`, :class:`BinTableHDU`",  "``PCOUNT``, ``GCOUNT``"
   ":class:`GroupsHDU`", "``NAXIS1``, ``GCOUNT``, ``PCOUNT``, ``GROUPS``"
   ":class:`TableHDU`, :class:`BinTableHDU`", "``TFIELDS``, ``TFORM``, ``TBCOL``"

> Structural keywords are checked and updated in 4 different places:

Reviewer 3: ```suggestion
Structural keywords are checked and updated in four different places:
```

1. When reading a file
2. When setting the associated data
3. When setting the :class:`Header` object
4. When writing a file

> This can often cause confusion because any hand-written values users might assign

Reviewer 3: ```suggestion
This can often cause confusion because any handwritten values users might assign
```

> to those keywords will be overwrittten. Consider the following examples:

Reviewer 4: Again I would remove the first part of the sentence and use "Any manually set values to these keywords ..."

.. code-block:: python

    >>> data = np.random.random((3,3))
    >>> ph = fits.PrimaryHDU(data)

    # The NAXIS keyword is automatically updated based on the data
    >>> ph.header['NAXIS']
    2

    # Changing the data will overwrite any changes the user makes
    >>> ph.header['NAXIS'] = 0
    >>> ph.header['NAXIS']
    0
    >>> data2 = np.random.random((3,3,3))
    >>> ph.data = data2
    >>> ph.header['NAXIS']
    3

    # Setting the Header object will also trigger a check
    >>> ph2 = fits.PrimaryHDU()
    >>> ph2.header['NAXIS']
    0
    >>> ph.header['NAXIS'] = 99
    >>> ph.header = ph2.header
    >>> ph.header['NAXIS']
    3

    # Finally, writing a file will trigger a check
    >>> ph.header['NAXIS'] = 99
    >>> ph.writeto('tmp.fits')
    >>> ph.header['NAXIS']
    3

.. warning::
  In other words, the structural keywords are a serialization detail that users

>   should almost never manipulate manually except in some rare corner cases.

Reviewer 4: I think the "almost never" is enough. It doesn't need the "except in some rare corner cases".

    See also :ref:`sphx_glr_generated_examples_io_modify-fits-header.py`.

> .. _structural_keywords:

Reviewer: I think this should go in another page, not in the main one which is already very long. Probably in http://docs.astropy.org/en/latest/io/fits/usage/headers.html#header-of-an-hdu ? As a subsection of "Header of an HDU" to emphasize the close relation between the header and its HDU ?
Then this section could be linked from the main page.

Author: I can see both sides of the argument here. This idea originally came from #3836, which, in my mind, was primarily concerned with getting this information in front of new users as quickly as possible. This is why I initially put this on the "Getting Started" page.

On the other hand, I think we're already getting close to "too detailed" for an introduction, so maybe it does make sense to really go into these details in a separate page.

So I guess I'm arguing for at least some mention of this to remain on the "Getting Started" page, but I'm fine adding more information somewhere else. The question then becomes what the minimal useful amount of information for the main page is. Maybe something as simple as a list of all structural keywords without any `*HDU` context and a warning not to touch them?

Structural Keywords
"""""""""""""""""""

FITS keywords mix up both metadata and critical information file structure that is
needed to parse the file. These *structural* keywords can be silently modified by

> :mod:`astropy.io.fits` and, in general, should not be touched by the user.

Reviewer: The important message is that these structural keywords should not be modified manually because the HDU/Header objects will manage the keywords. So I think that these attributes should be described together with the keywords. The table below is a good start, but it would be even better to tell which attribute is linked to a given keyword. E.g. NAXIS* is determined by `.data.shape`, BITPIX by `.data.dtype`, etc. Some keywords can just be ignored (SIMPLE, PCOUNT, GCOUNT). Table keywords are linked to the `Column` attributes (http://docs.astropy.org/en/latest/io/fits/usage/table.html#column-creation).

Author: I really like this idea. Once we decide where this table will go (see above) I'll add this additional information. Thanks for the link to the Table stuff.

Reviewer 5: I think the long-term goal should be to make the structural headers unsettable in normal use (I can imagine a need for doing it deeper in the implementation, or when constructing unusual FITS file formats).  But that seems like potentially a great deal of development effort, perhaps disproportionate to its value, so I do support as an interim measure (potentially one that might be around for years - let's be realistic) having "what not to do" documentation, perhaps of the form "In general, one should not attempt to set the headers in Table X directly, as they must remain consistent with the properties of the actual data payload of the file."  I think having the table mentioned above is an excellent idea in this context.

The specific set of structural keywords used by the FITS standard varies with HDU type.
The following table lists which keywords are associated with each HDU type:

.. csv-table:: Structural Keywords
   :header: "HDU Type", "Structural Keywords"
   :widths: 20, 20

   "All", "``SIMPLE``, ``BITPIX``, ``NAXIS``"
   ":class:`PrimaryHDU`", "``EXTEND``"
   ":class:`ImageHDU`, :class:`TableHDU`, :class:`BinTableHDU`",  "``PCOUNT``, ``GCOUNT``"
   ":class:`GroupsHDU`", "``NAXIS1``, ``GCOUNT``, ``PCOUNT``, ``GROUPS``"
   ":class:`TableHDU`, :class:`BinTableHDU`", "``TFIELDS``, ``TFORM``, ``TBCOL``"

Structural keywords are often checked and updated as a consequence of common operations.
For example, when:

1. Reading a file
2. Setting the associated data
3. Setting the :class:`Header` object
4. Writing a file

> 5. Calling an HDU's :func:`verify` method

Reviewer 4: That `verify` reference currently breaks the doc-build.
